### DISCUSSION TITLE
Rebuild the example of skfuzzy with fuzzylogic:

### DISCUSSION THREAD

**Discussion by born2bwater** [Python-Fuzzylogic/fuzzylogic#47] [Q&A]
Hi,
I am familiarizing myself with your super nice package. Thanks for making it publicly available.
During my search I came across this example: [Tipping Problem](https://pythonhosted.org/scikit-fuzzy/auto_examples/plot_tipping_problem.html)
The example has the result 20% tip.
With the code below I get: 2%

![image](https://github.com/Python-Fuzzylogic/fuzzylogic/assets/174632862/df77f0de-2c68-405e-8e47-ff2f23c31bc9)

I would expect that it is the combination of the rules that makes the difference. What would you recommend to do?

```
from fuzzylogic.classes import Domain, Set, Rule
from fuzzylogic.functions import R, S, alpha, trapezoid

res = 0.1

food = Domain("food_quality", 0, 10, res = res)
serv = Domain("service_quality", 0, 10, res = res)
tipa = Domain("tip_amount", 0, 25, res = res)

food.lo = S(0, 5)
food.md = alpha(floor = 0, ceiling = 1, func = trapezoid(0, 5, 5, 10))
food.hi = R(5, 10)

serv.lo = S(0, 5)
serv.md = alpha(floor = 0, ceiling = 1, func = trapezoid(0, 5, 5, 10))
serv.hi = R(5, 10)

tipa.lo = S(0, 13)
tipa.md = alpha(floor = 0, ceiling = 1, func = trapezoid(0, 13, 13, 25))
tipa.hi = R(13, 25)

rules = Rule({
    (food.hi, serv.hi): tipa.hi,
    (food.lo, serv.lo): tipa.lo,
    (serv.md, ): tipa.md,
})

values = {food: 6.5, serv: 9.8}
print(R1(values), R2(values), R3(values), rules(values))
```

**Comment by amogorkon** [comment]
Hi and thanks you!

The first thing I'm noticing is their use of trimf, which is the triangular function which should correspond to fuzzylogic.functions.triangular (at least one would rightfully expect so!).

**Comment by amogorkon** [comment]
The next thing I'm noticing is `x_qual = np.arange(0, 11, 1)` which corresponds to your `food = Domain("food_quality", 0, 10, res = res)` (changing resolution and domain may change the end result quite a bit since weights are shifted)

**Reply by amogorkon** [reply]
Please note that unlike them, I am using a customized arange function:
```
        if int(self._res) == self._res:
            return np.arange(self._low, self._high + self._res, int(self._res))
        else:
            return np.linspace(self._low, self._high, int((self._high - self._low) / self._res) + 1)
```
which makes it a bit fiddly to directly convert their example. 

this is because engineers who are working with physical sensor-ranges normally don't think in terms of math ranges.

**Reply by born2bwater** [reply]
hmm... I felt that way. Still it was helpful to work on another example besides your nice showcase

**Reply by amogorkon** [reply]
Next step would be to visually check if the plots match your expectations...

**Reply by amogorkon** [reply]
Ah, interesting. They have an fuzz.interp_membership function - smart. Let's steal that 😁 (probably just a wrapper for a np interpolation, but still)

**Reply by amogorkon** [reply]
You're using the Rule class to map the use of OR in their code? I'd immediately think of fuzzylogic.combinators.MAX when I read OR... or the `|` operator between fuzzy-functions, which maps to MAX.

**Reply by born2bwater** [reply]
> Next step would be to visually check if the plots match your expectations...

Food:
![image](https://github.com/Python-Fuzzylogic/fuzzylogic/assets/174632862/aa1b2032-9bc5-468b-84e0-c4b02b65b404)

Service
![image](https://github.com/Python-Fuzzylogic/fuzzylogic/assets/174632862/65719edc-5589-4ad4-91b1-3f012d8022ed)

Tip:
![image](https://github.com/Python-Fuzzylogic/fuzzylogic/assets/174632862/6efcff2f-6ebf-495a-bcfd-0ed4afd36285)

**Reply by amogorkon** [reply]
I find it a bit hard to read their mix of rules, plots and functions, tbh.. but here we go

**Comment by amogorkon** [comment]
I think I found the issue.

```
from fuzzylogic.classes import Domain, Rule
from fuzzylogic.functions import R, S, triangular

res = 1

food = Domain("food_quality", 0, 10, res=res)
serv = Domain("service_quality", 0, 10, res=res)
tip = Domain("tip_amount", 0, 25, res=res)


food.lo = S(0, 5)
food.md = triangular(0, 10)
food.hi = R(5, 10)

serv.lo = S(0, 5)
serv.md = triangular(0, 10)
serv.hi = R(5, 10)

tip.lo = S(0, 13)
tip.md = triangular(0, 25)
tip.hi = R(13, 25)

rules = Rule({
    (food.lo, serv.lo): tip.lo,
    (food.md, serv.md): tip.md,
    (food.hi, serv.hi): tip.hi,
})

values = {food: 6.5, serv: 9.8}
print(rules(values))
```
Returns `19.513574660633484` which is close to the expected 20% - it seems the resolution indeed was the culprit - which means the interpolation actually makes all the difference.

(Now looking at my code and comparing it to the original makes me wonder if I missed anything, it's so much more concise.. 🤔)

**Reply by amogorkon** [reply]
I actually find their code very hard to read because they mix fuzzylogic math with the underlying numpy functions, so you need to know exactly what you're doing - and pure fuzzylogic implication math already is tricky, imo. For instance you mustn't mix domains before the actual implication, which numpy doesn't care about.

**Reply by amogorkon** [reply]
I'll have to look at the scikit stuff step by step to see the actual differences..

**Reply by amogorkon** [reply]
Sigh. Their example doesn't work with 3.12, which I'm running..

**Reply by amogorkon** [reply]
Let's calculate by hand what the expected output should be..

**Reply by amogorkon** [reply]
Figuring out if https://pythonhosted.org/scikit-fuzzy/_modules/skfuzzy/defuzzify/defuzz.html#centroid actually does the same as my code is brain-wrecking 💩 
... feels like an exam question 😛  - at least it's an interesting exercise to get back into this part of the code. Time to take care of these rough edges.

**Reply by born2bwater** [reply]
Hey,
thank you very much for all these interesting answers. I had to work a bunch and couldnt answer right away. I'll update my example and try to get as close as you did with the result.
I couldnt use the scikit aswell, which is why I couldn't rebuild their example in their code.
Yeah I also think its a interesting exercise, though i am far from understanding everything you said... need some time I guess :-)

**Reply by amogorkon** [reply]
I'm not even sure if their approach of resolving cog works correctly. The idea of applying geometry is compelling, assuming the result is a simple geometric form - but why would they be? The moment you apply AND, OR, alpha or linguistic hedges you end up with non-trivial geometric forms... Their lack of inline documention on the actual algorithms doesn't help either 🤔

**Reply by amogorkon** [reply]
I rewrote the cog method:

```
                # iterate over the conditions and calculate the actual values and weights contributing to cog
                target_weights: list[tuple[Set, int | float]] = []
                target_domain = list(self.conditions.values())[0].domain
                for if_sets, then_set in self.conditions.items():
                    actual_values: list[int | float] = []
                    for s in if_sets:
                        actual_values.append(s(values[s.domain]))
                    x = min(actual_values, default=0)
                    if x > 0:
                        target_weights.append((then_set, x))
                if not target_weights:
                    return None
                sum_weights = 0
                sum_weighted_cogs = 0
                for then_set, weight in target_weights:
                    sum_weighted_cogs += then_set.center_of_gravity() * weight
                    sum_weights += weight
                index = sum_weighted_cogs / sum_weights
                return (target_domain._high - target_domain._low) / len(
                    target_domain.range
                ) * index + target_domain._low
```
I think that makes the whole thing much more approachable, don't you think?

**Reply by born2bwater** [reply]
I did get a similar result now and I didnt think that the resolution could change the outcome so drastically - lesson learned.
Just to make sure I understand correctly: your above method is used for the defuzzyfication. It calculates the point that is representing the area under the function the best. Hence it makes a difference which form the area of the function has. My naive question is (I am a geographer), if it is not easiest to calculate multiple cog's (for each single function connected by OR) and then calculate a cog of cogs (interpolate). I ask that because I cannot fully read your above code.

**Reply by amogorkon** [reply]
While, of course, I hope that my code is correct, it's possible that I made a mistake. That's why I rewrote it - to better scrutinize and improve.

My code does the following:
1. for every domain:value in the input calculate the membership values for all relevant fuzzy sets
2. for every IF-THEN rule take the min (AND) of the membership values and associate the THEN fuzzy set with the result as weight
3. calculate the weighted average for the cog of all THEN fuzzy sets as dimensionless index on the domain array
4. adjust for the limits and resolution of the domain to translate the index to the actual end value.

I guess it shows I never really cared much for defuzzification and put my emphasis on correctness and efficiency of the membership functions and combinators... 🤔

**Reply by born2bwater** [reply]
Hey,

thanks for the explanation. I cannot claim to being able to evaluate your code, but it is read-, and understandable. How it works in the bigger picture is yet beyond my scope. But I will look further into the options your package offers for me =D

**Reply by born2bwater** [reply]
Heya,

hope you're doing great.

I played around with your fuzzylogic package in a separate python environment (called fuzzy). Now that my main project is far enough to add fuzzy logic to it, i re-installed your package in a different env (called gis). The same example we posted above is not working in the gis environment. Neither does the new one. But it still works in the fuzzy.
Thus either there is a bad interaction betwenn different packages in the gis environment or you changed something in your code that causes that error (code see below). What do you think?
(PS: does github not offer to have collabseable sections in posts?)

```
from fuzzylogic.classes import Domain, Rule
from fuzzylogic.functions import R, S, triangular

res = 1

food = Domain("food_quality", 0, 10, res=res)
serv = Domain("service_quality", 0, 10, res=res)
tip = Domain("tip_amount", 0, 25, res=res)


food.lo = S(0, 5)
food.md = triangular(0, 10)
food.hi = R(5, 10)

serv.lo = S(0, 5)
serv.md = triangular(0, 10)
serv.hi = R(5, 10)

tip.lo = S(0, 13)
tip.md = triangular(0, 25)
tip.hi = R(13, 25)

rules = Rule({
    (food.lo, serv.lo): tip.lo,
    (food.md, serv.md): tip.md,
    (food.hi, serv.hi): tip.hi,
})

values = {food: 6.5, serv: 9.8}
print(rules(values))
ohalala
Traceback (most recent call last):

  Cell In[13], line 30
    print(rules(values))

  File ~\AppData\Local\miniconda3\envs\gis\Lib\site-packages\fuzzylogic\classes.py:447 in __call__
    index = sum(v.center_of_gravity * x for v, x in weights) / sum(x for v, x in weights)

  File ~\AppData\Local\miniconda3\envs\gis\Lib\site-packages\fuzzylogic\classes.py:447 in <genexpr>
    index = sum(v.center_of_gravity * x for v, x in weights) / sum(x for v, x in weights)

TypeError: unsupported operand type(s) for *: 'method' and 'float'
```

**Reply by amogorkon** [reply]
That's the old (fixed, but not released) bug. It's just missing a () at `v.center_of_gravity()`. I really should push a release. Possibly later today.

**Reply by born2bwater** [reply]
Oh nos. I am sorry I should have known... I once corrected it.
